test/jit/test_graph_rewrite_passes.py

```diff
@@ -4,58 +4,131 @@
+        self.check_op_presence([], check_not, torch._C._jit_pass_fuse_linear, model1)
+        model1(x1)  # make sure it runs
+
+        model2 = torch.jit.trace(FunctionalLinear(w1, None), [x1])
+        self.check_single_replacement("aten::matmul", "aten::linear", torch._C._jit_pass_fuse_linear, model2)
+        self.check_op_presence([], check_not, torch._C._jit_pass_fuse_linear, model2)
+        model2(x1)  # make sure it runs
 
         # check matmuls are not fused
-        class Matmul(torch.nn.Module):
-            def __init__(self, weight):
-                super(Matmul, self).__init__()
-                self.weight = weight
-
-            def forward(self, x):
-                return torch.matmul(x, self.weight)
-
-        x = torch.rand(5, 6, 5)
-        w = torch.rand(5, 5, 100)
-        model = torch.jit.trace(Matmul(w), [x])
-        torch._C._jit_pass_fuse_linear(model.graph)
-        # check 3d matmul is not fused
-        FileCheck().check("aten::matmul").run(model.graph)
-        FileCheck().check_not("aten::linear").run(model.graph)
-        # make sure it runs
-        model(x)
+        x3 = torch.rand(5, 6, 5)
+        w3 = torch.rand(5, 5, 100)
+        model3 = torch.jit.trace(FunctionalMatmul(w3), [x3])
+        check_not3 = ["aten::linear"]
+        self.check_single_replacement("aten::matmul", "aten::matmul", torch._C._jit_pass_fuse_linear, model3)
+        self.check_op_presence([], check_not3, torch._C._jit_pass_fuse_linear, model3)
+        model3(x3)  # make sure it runs
+
+    def test_vulkan_insert_pre_packed_ops(self):
+        x1 = torch.rand(3)
+        w1 = torch.rand(5, 3)
+        b1 = torch.rand(5)
+        model1 = torch.jit.trace(FunctionalLinear(w1, b1), [x1])
+        check_not1 = ["aten::matmul", "aten::add_", "aten::t"]
+        self.check_single_replacement(
+            "aten::matmul", 
+            "vulkan_prepack::linear_run", 
+            torch._C._jit_pass_vulkan_insert_prepacked_ops, 
+            model1
+        )
+        self.check_op_presence([], check_not1, torch._C._jit_pass_vulkan_insert_prepacked_ops, model1)
+        model1(x1)  # make sure it runs
+
+        conv2d_in_channels = 3
+        conv2d_out_channels = 4
+        conv2d_kernel = 2
+        conv2d_weight = torch.rand(
+            conv2d_out_channels,
+            conv2d_in_channels,
+            conv2d_kernel,
+            conv2d_kernel,
+        )
+        conv2d_bias = torch.rand(conv2d_out_channels)
+        x2_shape = (3, 2, 5)
+        x2 = torch.rand(x2_shape)
+        model2 = torch.jit.trace(FunctionalConv2d(conv2d_weight, conv2d_bias), [x2])
+        check_not2 = ["aten::_convolution"]
+        check_yes2 = ["vulkan_prepack::conv2d_clamp_run", "vulkan_prepack::conv2d_clamp_prepack"]
+        self.check_single_replacement(
+            "aten::_convolution", 
+            "prim::Constant", 
+            torch._C._jit_pass_vulkan_insert_prepacked_ops, 
+            model2
+        )
```

Comment:
This is the location if the concern I mentioned privately. I expected the following test to pass, but it does not:
```
self.check_single_replacement("aten::_convolution",  "vulkan_prepack::conv2d_clamp_run", torch._C._jit_pass_vulkan_insert_prepacked_ops, model2)
```
To debug, I inspected the `sourceRange()` of the node of type `vulkan_prepack::conv2d_clamp_run` and got what looks like an error and not the expected `sourceRange()` content:
```
           %transposed:bool, %output_padding:int[], %groups:int, %benchmark:bool,
          %deterministic:bool, %cudnn_enabled:bool, %allow_tf32:bool):
        %r = aten::conv2d(%a, %w, %b, %stride, %padding, %dilation, %groups)
             ~~~~ <--- HERE
        return (%r)
```
@kimishpatel @SS-JIA I'd appreciate any guidance on this issue.

Reply:
What is the exact output of the test? What is the sourceRange of the `aten::_convolution`?

Reply:
The vulkan rewrite pass seems to expect `aten::conv2d`, do you know what the differences are between this op and `aten::_convolution`?

Reply:
`_convolution` I think can do 1d and 3d convs. And eventually all conv1d, 2d and 3d go to _convolution.

Reply:
Spent quite a bit experimenting. Here's what I found:
> What is the exact output of the test?
```
======================================================================
FAIL: test_vulkan_insert_pre_packed_ops (jit.test_graph_rewrite_passes.TestGraphRewritePasses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/hagiopol/pytorch/test/jit/test_graph_rewrite_passes.py", line 143, in test_vulkan_insert_pre_packed_ops
    self.check_single_replacement(
  File "/Users/hagiopol/pytorch/test/jit/test_graph_rewrite_passes.py", line 61, in check_single_replacement
    self.assertTrue(source_range_1 == source_range_2)
AssertionError: False is not true
```
> What is the sourceRange of the aten::_convolution?
```
/Users/hagiopol/pytorch/test/jit/test_graph_rewrite_passes.py(16): forward
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/nn/modules/module.py(1098): _slow_forward
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/nn/modules/module.py(1110): _call_impl
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/jit/_trace.py(958): trace_module
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/jit/_trace.py(741): trace
/Users/hagiopol/pytorch/test/jit/test_graph_rewrite_passes.py(143): test_vulkan_insert_pre_packed_ops
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/case.py(550): _callTestMethod
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/case.py(592): run
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py(1574): _run_with_retry
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py(1604): run
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/case.py(651): __call__
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/suite.py(122): run
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/suite.py(84): __call__
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/suite.py(122): run
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/suite.py(84): __call__
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/runner.py(184): run
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/main.py(271): runTests
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/unittest/main.py(101): __init__
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py(652): run_tests
/Users/hagiopol/pytorch/test/test_jit.py(16425): <module>
```
>The vulkan rewrite pass seems to expect `aten::conv2d`
This is what I expected as well, but upon inspection of the node kinds generated by `torch.nn.functional.conv2d` I see only `aten::_convolution`.

Reply:
> This is what I expected as well, but upon inspection of the node kinds generated by torch.nn.functional.conv2d I see only aten::_convolution.

This is due to the difference between `torch.jit.trace` vs. `torch.jit.script`. I can explain more about the differences later.

Reply:
Ok so if you do torch.jit.script I expect the test to pass. Here is the reason why this test fails:

When you do tracing you get `at::_convolution`. Rewrite passes are written in a way to do preprocessing to replace few common patterns with canonical one which then graph rewrite relies on. In this case graph rewrite will always rewrite `at::conv2d` but conv2d might be expressed differently like via `_convolution`. So first thing we do is to replace `_convolution` with conv1d/2d/3d etc.
See here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/graph_rewrite_helper.cpp#L56-L223
These passes dont preserve the source range info and hence we are not getting what we expect. If this one is also fixed similarly, not part of this task, we should have the test pass.